File: windows/net/samples/mux_mp4_remux_file/Options.cs
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

using System;
using System.IO;
using CommandLine;
using CommandLine.Text;

/*
command line options:

	mux_mp4_remux_file --input audio.mp4 video.mp4 --output output.mp4 [--fast-start]
*/

namespace MuxMp4RemuxFileSample
{
    class Options
    {
        // mux_mp4_avc_aac_file options
        [Option('?', "help", HelpText = "Display this help screen")]
        public bool Help { get; set; }

        [Option('i', "input", HelpText = "input file (mp4)")]
        public string InputFile { get; set; }

        [Option('o', "output", HelpText = "output file")]
        public string OutputFile { get; set; }

        [Option('f', "fast-start", HelpText = "mp4 fast start")]
        public bool FastStart { get; set; }

        // The program parses the command line options and sets these properties
        public bool Error { get; private set; }

        string GetUsage()
        {
            return HelpText.AutoBuild(this, (HelpText current) => HelpText.DefaultParsingErrorsHandler(this, current));
        }

        void PrintUsage()
        {
            Console.WriteLine("Usage: mux_mp4_remux_file --input file.mp4 --output output.mp4 [--fast-start]");
            Console.WriteLine(GetUsage());
        }

        void ResetOptions()
        {
            Help = false;
            Error = false;
            InputFile = null;
            OutputFile = null;
        }

        void SetDefaultOptions()
        {
            string exeDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);

            InputFile = Path.Combine(exeDir, "..\\assets\\mov\\big_buck_bunny_trailer_iphone.m4v");
            OutputFile = Path.Combine(exeDir, "mux_mp4_remux_file.mp4");

            Console.WriteLine("Using default options: --input " + InputFile + " --output " + OutputFile);
        }

        public bool Prepare(string[] args)
        {
            ResetOptions();

            if (args.Length == 0)
            {
                SetDefaultOptions();
            }
            else
            {
                if (!CommandLine.Parser.Default.ParseArguments(args, this))
                {
                    Console.WriteLine("Syntax error");
                    PrintUsage();
                    Error = true;
                    return false;
                }
            }

            if (Help)
            {
                PrintUsage();
                return false;
            }

            if (!Validate())
            {
                Console.WriteLine("\nRequired options are not set!");
                PrintUsage();
                Error = true;
                return false;
            }

            return true;
        }

        bool Validate()
        {
            bool res = true;

            Console.Write("Input file: ");
            if (!string.IsNullOrEmpty(InputFile))
            {
                Console.WriteLine(InputFile);
            }
            else
            {
                Console.WriteLine("[not set]");
                res = false;
            }

            Console.Write("Output file: ");
            if (string.IsNullOrEmpty(OutputFile))
            {
                Console.WriteLine("[not set]");
                res = false;
            }
            else
            {
                Console.WriteLine(OutputFile);
            }

            Console.WriteLine("Fast-start: " + (FastStart ? "yes" : "no"));

            return res;
        }
    }
}

File: windows/cpp/samples/mux_mp4_remux_file/options.cpp
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

#include "stdafx.h"
#include "options.h"
#include "util.h"
#include <atlstr.h>
#include "program_options.h"

using namespace std;
using namespace primo::avblocks;	
using namespace primo::codecs;	
using namespace primo::program_options;

void help(OptionsConfig<wchar_t>& optionsConfig)
{
    wcout << endl;
    wcout << L"Usage: mux_mp4_remux_file --input file.mp4 --output output.mp4 [--fast-start]" << endl;
    doHelp(wcout, optionsConfig);
}

bool validateOptions(Options& opt)
{    
    if (opt.input_file.empty())
    {
        return false;
    }

    if (opt.output_file.empty())
    {
        return false;
    }
    return true;
}

void setDefaultOptions(Options& opt)
{
    if(opt.input_file.empty())
        opt.input_file = getExeDir() + L"\\..\\assets\\mov\\big_buck_bunny_trailer_iphone.m4v";
    
    if(opt.output_file.empty())
        opt.output_file = getExeDir() + L"\\mux_mp4_remux_file.mp4";
}

ErrorCodes prepareOptions(Options& opt, int argc, wchar_t* argv[])
{
    if (argc < 2)
    {
        setDefaultOptions(opt);
        wcout << L"Using defaults:\n";
        wcout << L" --input " << opt.input_file;
        wcout << L" --output " << opt.output_file;
        wcout << endl;
        return Parsed;
    }

    OptionsConfig<wchar_t> optionsConfig;
    optionsConfig.addOptions()
        (L"help,?",			opt.help,						 L"")
        (L"fast_start,f",	opt.fast_start,					 L"mp4 fast start")
        (L"input,i",		opt.input_file,	    wstring(),	 L"input file (mp4)")
        (L"output,o",		opt.output_file,	wstring(),	 L"output file");

    try
    {
        scanArgv(optionsConfig, argc, argv);
    }
    catch (ParseFailure<wchar_t> &ex)
    {
        wcout << ex.message() << endl;
        help(optionsConfig);
        return Error;
    }

    if(opt.help)
    {
        help(optionsConfig);
        return Command;
    }

    if (!validateOptions(opt))
    {
        help(optionsConfig);
        return Error;
    }

    return Parsed;
}
